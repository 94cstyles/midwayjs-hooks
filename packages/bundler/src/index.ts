import omit from 'lodash/omit'
import { createUnplugin, UnpluginOptions } from 'unplugin'
import {
  AbstractRouter,
  ApiRoute,
  EXPORT_DEFAULT_FUNCTION_ALIAS,
  parseApiModule,
} from '@midwayjs/hooks-core'
import jiti from 'jiti'

jiti().register()

interface BundlerConfig {
  name: string
  router: AbstractRouter
}

export abstract class AbstractBundlerAdapter {
  constructor(private config: BundlerConfig) {}

  getName() {
    return this.config.name
  }

  getRouter() {
    return this.config.router
  }

  getUnplugin(): Partial<UnpluginOptions> {
    return {}
  }

  transformApiRoutes(apis: ApiRoute[]) {
    return apis
  }

  generateClient(apis: ApiRoute[]) {
    let fetcherId = 0
    const importClient = new Map<string, [string, string, number]>()

    const importCodes = new Set<string>()
    const functionCodes = []

    for (const {
      trigger,
      functionName,
      functionId,
      useInputMetadata,
    } of apis) {
      if (!trigger.requestClient) continue
      const { fetcher, client } = trigger.requestClient

      const clientId = `${client}_${fetcher}`
      if (!importClient.has(clientId)) {
        importClient.set(clientId, [client, fetcher, fetcherId++])
      }

      const [clientName, fetcherName, idx] = importClient.get(clientId)
      const uniqueFetcher = `rpc$${idx}`
      importCodes.add(
        `import { ${fetcherName} as ${uniqueFetcher} } from '${clientName}'`
      )

      const exportModifier =
        functionName === EXPORT_DEFAULT_FUNCTION_ALIAS ? 'default' : ''
      functionCodes.push(`
        export ${exportModifier} function ${functionName} (...args) {
          const route = ${functionName}.route;
          return ${uniqueFetcher}(...args, route)
        };
        ${functionName}.route = {
          trigger: ${JSON.stringify(omit(trigger, 'requestClient'))},
          functionId: '${functionId}',
          useInputMetadata: ${!!useInputMetadata},
        };
      `)
    }

    const code = [
      '// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten.',
      ...importCodes,
      ...functionCodes,
    ].join('\n')
    return this.formatCode(code)
  }

  private formatCode(code: string) {
    try {
      const prettier = require('prettier')
      return prettier.format(code, {
        semi: true,
        singleQuote: true,
        parser: 'babel',
      })
    } catch (e) {
      console.log('format code error', e)
      return code
    }
  }
}

export function createBundlerPlugin(adapter: AbstractBundlerAdapter) {
  return createUnplugin((options) => {
    return {
      name: adapter.getName(),
      enforce: 'pre',
      transformInclude(id: string) {
        return adapter.getRouter().isApiFile(id)
      },
      async transform(code: string, id: string) {
        const apis = parseApiModule(require(id), id, adapter.getRouter())
        const transformedApis = adapter.transformApiRoutes(apis)
        return adapter.generateClient(transformedApis)
      },
      ...adapter.getUnplugin(),
    }
  })
}
