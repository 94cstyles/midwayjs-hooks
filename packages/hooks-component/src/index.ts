import { createConfiguration, IMidwayContainer } from '@midwayjs/core'

import { Inject, Controller, Get, Post, Provide } from '@midwayjs/decorator'

import { als } from '@midwayjs/hooks/lib/api/async_hooks/als'
import { EnhancedFunc } from '@midwayjs/hooks/lib/hoc/type'
import {
  WebRouterConfig,
  WebRouter,
  getFunctionId,
} from '@midwayjs/hooks-router'

interface HooksConfig extends Omit<WebRouterConfig, 'source'> {}

export const fnMap = new Map<EnhancedFunc, any>()

export const createHooks = (config: HooksConfig) => {
  const configuration = createConfiguration({
    namespace: 'hooks',
    directoryResolveFilter: createResolveFilter(config),
  })
    .onReady(() => {})
    .onStop(() => {})

  return {
    Configuration: configuration,
  }
}

function createResolveFilter(config: HooksConfig) {
  return config.routes.map((route) => {
    return {
      pattern: route.baseDir,
      filter: (
        mod: {
          default?: any
          [key: string]: any
        },
        sourceFilePath: string,
        container: IMidwayContainer
      ) => {
        const root: string = container.get('appDir')
        const router = new WebRouter(root, {
          routes: config.routes,
          source: (config as any).source,
        })

        // export default
        const defaultExports = mod.default || mod
        if (typeof defaultExports === 'function') {
          createFunctionContainer({
            container,
            router,
            fn: defaultExports,
            sourceFilePath,
            isExportDefault: true,
          })
        }

        for (const key of Object.keys(mod)) {
          if (key === 'default') {
            continue
          }

          const value = mod[key]
          if (typeof value === 'function') {
            createFunctionContainer({
              container,
              router,
              fn: value,
              sourceFilePath,
              isExportDefault: false,
            })
          }
        }
      },
    }
  })
}

function createFunctionContainer(config: {
  container: IMidwayContainer
  fn: EnhancedFunc
  sourceFilePath: string
  router: WebRouter
  isExportDefault: boolean
}) {
  const { container, fn, router, sourceFilePath, isExportDefault } = config

  const fnName = isExportDefault ? '$default' : fn.name
  const id = getFunctionId({
    router,
    sourceFilePath,
    functionName: fnName,
    isExportDefault,
  })

  const containerId = 'hooks_func::' + id
  const httpPath = router.getHTTPPath(sourceFilePath, fnName, isExportDefault)
  const HttpMethod = fn.length === 0 ? Get : Post

  fnMap.set(fn, {
    containerId,
    httpPath,
    httpMethod: fn.length === 0 ? 'GET' : 'POST',
  })

  @Provide(containerId)
  @Controller('/')
  class FunctionContainer {
    @Inject()
    ctx: any

    @HttpMethod(httpPath, { middleware: fn.middleware || [] })
    async handler() {
      const bindCtx = {
        ctx: this.ctx,
      }

      let args = this.ctx?.request?.body?.args || []
      if (typeof args === 'string') {
        args = JSON.parse(args)
      }

      return await als.run(bindCtx, () => fn(...args))
    }
  }

  container.bind(containerId, FunctionContainer)
}
