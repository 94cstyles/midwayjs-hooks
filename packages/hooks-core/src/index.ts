export * from './router'
export * from './component'
export * from './runtime'
export * from './request'
export * from './request/generate'
export * from './config'
export * from './proxy'

export * from './types/common'
export * from './types/http'
export * from './types/serverless'
export * from './types/config'
