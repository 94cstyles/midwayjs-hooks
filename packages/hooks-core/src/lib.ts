import superjson from 'superjson'
import consola from 'consola'

export { superjson, consola }
