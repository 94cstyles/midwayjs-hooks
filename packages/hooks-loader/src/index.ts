import { LambdaHTTPMethod } from '@midwayjs/hooks-shared'
import { codeFrameColumns } from '@babel/code-frame'
import * as parser from '@babel/parser'
import traverse, { NodePath } from '@babel/traverse'
import * as t from '@babel/types'
import { Identifier, FunctionDeclaration } from '@babel/types'
import { loader } from 'webpack'
import { buildRequest, RenderParam } from './render'
import { helper, getFunctionHandlerName, BuiltinEnhancer } from '@midwayjs/next-hooks-compiler'
import { debug } from './util'

export default async function loader(this: loader.LoaderContext, source: string) {
  const callback = this.async()
  const resourcePath = this.resourcePath

  if (resourcePath.includes('node_modules')) {
    return callback(null, source)
  }

  const root = this.rootContext || (this as any).options?.context
  helper.root = root

  if (!helper.isLambdaFile(resourcePath)) {
    return callback(null, source)
  }

  debug('compile %s', resourcePath)

  let ast: t.File
  try {
    ast = parser.parse(source, {
      sourceType: 'module',
      plugins: [
        'typescript',
        'classProperties',
        'objectRestSpread',
        'optionalCatchBinding',
        'dynamicImport',
        'decorators-legacy',
        'asyncGenerators',
        'exportDefaultFrom',
        'exportNamespaceFrom',
        'optionalCatchBinding',
        'throwExpressions',
        'optionalChaining',
        'nullishCoalescingOperator',
      ],
    })
  } catch (error) {
    error.message = codeFrameColumns(source, error.loc, {
      highlightCode: true,
      message: error.message,
    })
    throw error
  }

  const funcs: RenderParam[] = []

  let err: Error

  /**
   * Cases in which export is allowed
   *
   * 1. export async function demo () {}
   * 2. export const demo = async function () {}
   * 3. export const demo = async () => {}
   * 4. export default async function demo () {}
   * 5. export default async function () {}
   * 6. export default async () => {}
   */
  traverse(ast, {
    ExportNamedDeclaration(path) {
      const func: RenderParam = {}
      const declaration = path.node.declaration

      if (t.isTSTypeAliasDeclaration(declaration) || t.isTSInterfaceDeclaration(declaration)) {
        return
      }

      if (t.isFunctionDeclaration(declaration)) {
        const id = declaration.id

        func.url = helper.getHTTPPath(resourcePath, id.name, false)

        const { method, params } = parseFunctionParams(declaration.params)
        func.method = method

        func.meta = {
          functionName: getFunctionHandlerName({
            sourceFilePath: resourcePath,
            isExportDefault: false,
            functionName: id.name,
          }),
          unstable_params: params,
        }
        func.functionId = id.name

        funcs.push(func)
        return
      }

      if (t.isVariableDeclaration(declaration)) {
        for (const variableDeclarator of declaration.declarations) {
          const id = variableDeclarator.id as Identifier
          const init = variableDeclarator.init

          func.url = helper.getHTTPPath(resourcePath, id.name, false)
          func.meta = {
            functionName: getFunctionHandlerName({
              sourceFilePath: resourcePath,
              isExportDefault: false,
              functionName: id.name,
            }),
          }
          func.functionId = id.name

          if (t.isArrowFunctionExpression(init) || t.isFunctionExpression(init)) {
            const { method, params } = parseFunctionParams(init.params)
            func.method = method
            func.meta.unstable_params = params
          } else if (t.isCallExpression(init) && BuiltinEnhancer.includes((init.callee as Identifier).name)) {
            const callback = init.arguments[1] as t.FunctionExpression
            const { method, params } = parseFunctionParams(callback.params)
            func.method = method
            func.meta.unstable_params = params
          } else {
            err = buildCodeFrameError(path, getErrorMessage())
            path.stop()
            return
          }

          funcs.push(func)
        }
        return
      }

      err = buildCodeFrameError(path, getErrorMessage())
      path.stop()
    },
    ExportDefaultDeclaration(path) {
      const declaration = path.node.declaration
      const func: RenderParam = {}

      let funcExpression: t.FunctionDeclaration | t.FunctionExpression | t.ArrowFunctionExpression = null

      if (t.isFunctionDeclaration(declaration) || t.isArrowFunctionExpression(declaration)) {
        funcExpression = declaration
      } else if (t.isCallExpression(declaration) && BuiltinEnhancer.includes((declaration.callee as Identifier).name)) {
        funcExpression = declaration.arguments[1] as t.FunctionExpression
      }

      if (!funcExpression) {
        err = buildCodeFrameError(path, getErrorMessage())
        path.stop()
        return
      }

      const functionName = getFunctionHandlerName({
        sourceFilePath: resourcePath,
        isExportDefault: true,
        functionName: getFunctionName(path),
      })

      const { method, params } = parseFunctionParams(funcExpression.params)
      func.method = method
      func.functionId = getFunctionName(path)
      func.isExportDefault = true
      func.url = helper.getHTTPPath(resourcePath, functionName, true)
      func.meta = {
        functionName,
        unstable_params: params,
      }

      funcs.push(func)
    },
  })

  if (err) {
    callback(err)
    return
  }

  const code = buildRequest(funcs, root, this.query)
  debug('compile %s', resourcePath)
  callback(null, code)
}

function getFunctionName(path: NodePath<t.ExportDefaultDeclaration>) {
  const declaration = path.node.declaration
  const defaultExportName = '$default'

  if (t.isFunctionDeclaration(declaration)) {
    return declaration?.id?.name || defaultExportName
  }

  return defaultExportName
}

function parseFunctionParams(params: FunctionDeclaration['params']) {
  const keys = []

  for (const param of params) {
    // (a) => {}
    if (t.isIdentifier(param)) {
      keys.push(param.name)
      continue
    }

    //  (a = 1) => {}
    if (t.isAssignmentPattern(param) && t.isIdentifier(param.left)) {
      keys.push(param.left.name)
      continue
    }

    // (...a) => {}
    if (t.isRestElement(param) && t.isIdentifier(param.argument)) {
      keys.push(param.argument.name)
      continue
    }

    console.log('Function parameter parsing error, node type: %s', param.type)
  }

  return {
    method: params?.length > 0 ? 'POST' : ('GET' as LambdaHTTPMethod),
    params: keys,
  }
}

function getErrorMessage() {
  return `
    When a function file is used as an API, only the following export methods are allowed
      1. export async function demo () {}
      2. export const demo = async function () {}
      3. export const demo = async () => {}
      4. export default async function demo () {}
      5. export default async function () {}
      6. export default async () => {}
  `
}

function buildCodeFrameError(path: NodePath<any>, message: string) {
  const root = path.findParent((p) => !p.parentPath)
  const frame = codeFrameColumns(
    root.toString(),
    {
      end: path.node.loc.end,
    },
    {
      message: message,
    }
  )

  return new Error(frame)
}
