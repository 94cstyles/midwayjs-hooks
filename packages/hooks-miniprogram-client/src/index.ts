import type { ApiBaseParam } from '@midwayjs/hooks-core'
import { isWeChatMiniProgram } from '@uni/env'

export const request = async (param: ApiBaseParam) => {
  if (isWeChatMiniProgram) {
    const { result } = await wx.cloud.callFunction({
      name: param.functionId,
      data: param.data,
    })
    return result
  }
  throw new Error('current miniprogram is not support')
}
