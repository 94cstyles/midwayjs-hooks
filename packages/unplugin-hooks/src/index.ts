import omit from 'lodash/omit'
import { createUnplugin, UnpluginOptions } from 'unplugin'
import {
  AbstractRouter,
  ApiRoute,
  EXPORT_DEFAULT_FUNCTION_ALIAS,
  loadApiRoutesFromFile,
} from '@midwayjs/hooks-core'

interface BundlerConfig {
  name: string
  router: AbstractRouter
  unplugin?: UnpluginOptions
}

export abstract class AbstractBundlerAdapter {
  constructor(private config: BundlerConfig) {}

  getName() {
    return this.config.name
  }

  getRouter() {
    return this.config.router
  }

  getUnplugin() {
    return this.config.unplugin || {}
  }

  transformApiRoutes(apis: ApiRoute[]) {
    return apis
  }

  generateClient(apis: ApiRoute[]) {
    let fetcherId = 0
    const importClient = new Map<string, [string, string, number]>()

    const importCodes = new Set<string>()
    const functionCodes = []

    for (const {
      trigger,
      functionName,
      functionId,
      hasMetadataInput,
    } of apis) {
      if (!trigger.requestClient) continue
      const { fetcher, client } = trigger.requestClient

      const clientId = `${client}_${fetcher}`
      if (!importClient.has(clientId)) {
        importClient.set(clientId, [client, fetcher, fetcherId++])
      }

      const [clientName, fetcherName, idx] = importClient.get(clientId)
      const uniqueFetcher = `rpc$${idx}`
      importCodes.add(
        `import { ${fetcherName} as ${uniqueFetcher} } from '${clientName}'`
      )

      const exportModifier =
        functionName === EXPORT_DEFAULT_FUNCTION_ALIAS ? 'default' : ''
      functionCodes.push(`
        export ${exportModifier} function ${functionName} (...args) {
          const route = {
            trigger: ${JSON.stringify(omit(trigger, 'requestClient'))},
            functionId: '${functionId}',
            hasMetadataInput: ${!!hasMetadataInput},
          }
          return ${uniqueFetcher}(route, ...args)
        }
      `)
    }

    const code = [...importCodes, ...functionCodes].join('\n')
    return this.formatCode(code)
  }

  private formatCode(code: string) {
    try {
      const prettier = require('prettier')
      return prettier.format(code, {
        semi: true,
        singleQuote: true,
        parser: 'babel',
      })
    } catch (e) {
      console.log('format code error', e)
      return code
    }
  }
}

export function createBundlerPlugin(adapter: AbstractBundlerAdapter) {
  return createUnplugin((options) => {
    return {
      name: adapter.getName(),
      enforce: 'pre',
      transformInclude(id: string) {
        return adapter.getRouter().isApiFile(id)
      },
      async transform(code: string, id: string) {
        const apis = loadApiRoutesFromFile(require(id), id, adapter.getRouter())
        const transformedApis = adapter.transformApiRoutes(apis)
        return adapter.generateClient(transformedApis)
      },
      ...adapter.getUnplugin(),
    }
  })
}
