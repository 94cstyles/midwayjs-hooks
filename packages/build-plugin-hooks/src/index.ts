import { IPlugin } from '@alib/build-scripts'
import { HooksWebpackPlugin } from '@midwayjs/unplugin-hooks'

const buildHooksRequest: IPlugin = (pluginApi) => {
  pluginApi.onGetWebpackConfig((config) => {
    config.plugin('@midwayjs/hooks').use(HooksWebpackPlugin())
  })
}

export default buildHooksRequest
