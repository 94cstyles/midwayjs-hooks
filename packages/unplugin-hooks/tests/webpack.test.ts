import { wrap } from 'jest-snapshot-serializer-raw'
import Memoryfs from 'memory-fs'
import { resolve as pathResolve } from 'path'

import webpack from 'webpack'

import { HooksWebpackPlugin } from '../src'

describe.skip('unplugin-hooks webpack', () => {
  const root = pathResolve(__dirname, './fixtures/base-app')
  const cwd = process.cwd()

  const getEntry = (path: string) => {
    return pathResolve(root, 'src/', path)
  }

  beforeEach(() => {
    process.chdir(root)
  })

  afterEach(() => {
    process.chdir(cwd)
  })

  test('Compile render', async () => {
    const output = await compile(root, getEntry('render/[...index].ts'))
    expect(wrap(output)).toMatchSnapshot()
  })

  test('Compile lambda', async () => {
    const output = await compile(root, getEntry('lambda/index.ts'))
    expect(wrap(output)).toMatchSnapshot()
  })

  test('Compile event', async () => {
    const output = await compile(root, getEntry('wechat/index.ts'))
    expect(wrap(output)).toMatchSnapshot()
  })

  test('the second build should match the first.', async () => {
    const first = await compile(root, getEntry('lambda/index.ts'))
    const second = await compile(root, getEntry('lambda/index.ts'))
    expect(first).toEqual(second)
  })

  test('non-lambda files should not be compiled', async () => {
    const output = await compile(root, getEntry('util/util.ts'))
    expect(wrap(output)).toMatchSnapshot()
  })
})

function compile(context: string, entry: string): Promise<string> {
  return new Promise((resolve) => {
    const compiler = webpack({
      context,
      entry,
      plugins: [HooksWebpackPlugin()],
      output: {
        path: '/',
        filename: 'client.js',
      },
      externals: {
        '@midwayjs/hooks/request': '@midwayjs/hooks/request',
        axios: 'axios',
      },
      mode: 'development',
      module: {
        rules: [
          {
            test: /\.tsx?$/,
            use: [
              {
                loader: 'ts-loader',
                options: {
                  transpileOnly: true,
                },
              },
            ],
          },
        ],
      },
    })

    compiler.outputFileSystem = new Memoryfs()

    compiler.run((err: any, stats) => {
      if (err) {
        console.error(err.stack || err)
        if (err.details) {
          console.error(err.details)
        }
        return
      }

      const info = stats.toJson()

      if (stats.hasErrors()) {
        console.error(info.errors)
      }

      if (stats.hasWarnings()) {
        console.warn(info.warnings)
      }

      const module = info.modules.find((mod) => mod.identifier.includes(entry))
      resolve(module.source)
    })
  })
}
