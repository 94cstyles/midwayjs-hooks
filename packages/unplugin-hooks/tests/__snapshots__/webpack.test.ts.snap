// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unplugin-hooks webpack Compile event 1`] = `
import { request as request$0 } from '@midwayjs/hooks-miniprogram-client';
export default function $default(...args) {
    const options = Object.assign({
        functionId: 'wechat-index',
        data: { args },
    }, {});
    return request$0(options);
}

`;

exports[`unplugin-hooks webpack Compile lambda 1`] = `
import { request as request$0 } from '@midwayjs/hooks/request';
export default function $default(...args) {
    const options = Object.assign({
        functionId: 'lambda-index',
        data: { args },
    }, {
        url: '/api',
        meta: {},
    });
    return request$0(options);
}

`;

exports[`unplugin-hooks webpack Compile render 1`] = `
import { request as request$0 } from '@midwayjs/hooks/request';
export default function $default(...args) {
    const options = Object.assign({
        functionId: 'render-index',
        data: { args },
    }, {
        url: '/',
        meta: {},
    });
    return request$0(options);
}
export function foo(...args) {
    const options = Object.assign({
        functionId: 'render-index-foo',
        data: { args },
    }, {
        url: '/foo',
        meta: {},
    });
    return request$0(options);
}
export function bar(...args) {
    const options = Object.assign({
        functionId: 'render-index-bar',
        data: { args },
    }, {
        url: '/bar',
        meta: {},
    });
    return request$0(options);
}

`;

exports[`unplugin-hooks webpack non-lambda files should not be compiled 1`] = `
export function isTrue(value) {
    return value === 'true' || value === true;
}

`;
